pkg/daemon/ceph/client/pool.go

```diff
@@ -418,6 +421,78 @@ func CreateReplicatedPoolForApp(context *clusterd.Context, clusterInfo *ClusterI
 	}
 
 	logger.Infof("creating replicated pool %s succeeded", poolName)
+
+	if checkFailureDomain {
+		if err = ensureFailureDomain(context, clusterInfo, clusterSpec, poolName, pool); err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+func ensureFailureDomain(context *clusterd.Context, clusterInfo *ClusterInfo, clusterSpec *cephv1.ClusterSpec, poolName string, pool cephv1.PoolSpec) error {
+	if pool.FailureDomain == "" {
+		logger.Debugf("skipping check for failure domain on pool %q as it is not specified", poolName)
+		return nil
+	}
+
+	logger.Debugf("checking that pool %q has the failure domain %q", poolName, pool.FailureDomain)
+	details, err := GetPoolDetails(context, clusterInfo, poolName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get pool %q details", poolName)
+	}
+
+	// Find the failure domain for the current crush rule
+	rule, err := getCrushRule(context, clusterInfo, details.CrushRule)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get crush rule %q", details.CrushRule)
+	}
+	currentFailureDomain := extractFailureDomain(rule)
+	if currentFailureDomain == pool.FailureDomain {
+		logger.Debugf("pool %q has the expected failure domain %q", poolName, pool.FailureDomain)
+		return nil
+	}
+	if currentFailureDomain == "" {
+		logger.Warningf("failure domain not found for crush rule %q, proceeding to create a new crush rule", details.CrushRule)
+	}
+
+	// Use a crush rule name that is unique to the desired failure domain
+	crushRuleName := fmt.Sprintf("%s_%s", poolName, pool.FailureDomain)
+	logger.Infof("updating pool %q failure domain from %q to %q with new crush rule %q", poolName, currentFailureDomain, pool.FailureDomain, crushRuleName)
+	logger.Infof("crush rule %q will no longer be used by pool %q", details.CrushRule, poolName)
+
+	// Create a new crush rule for the expected failure domain
+	if err := createReplicationCrushRule(context, clusterInfo, clusterSpec, crushRuleName, pool); err != nil {
+		return errors.Wrapf(err, "failed to create replicated crush rule %q", crushRuleName)
+	}
+
+	// Update the crush rule on the pool
+	if err := setCrushRule(context, clusterInfo, poolName, crushRuleName); err != nil {
+		return errors.Wrapf(err, "failed to set crush rule on pool %q", poolName)
+	}
+
+	logger.Infof("Successfully updated pool %q failure domain to %q", poolName, pool.FailureDomain)
+	return nil
+}
+
+func extractFailureDomain(rule ruleSpec) string {
+	// find the failure domain in the crush rule, which is the first step where the
+	// "type" property is set
+	for _, step := range rule.Steps {
```

Comment:
I like this approach better. Can we use `sets.NewString()` with `Has()` instead for this loop?

Reply:
Actually, this might or might not work. If we have a 2 steps rule or more we might find the step but it could be in a different rack. Maybe what we had earlier was stricter and we should stick to it, if the rule steps go deeper we should fail. This would mean the rule was not created by Rook and modified. So I think we should go back to using `rule.Steps[1]` with a length check.